### DISCUSSION TITLE
How to retrieve the names of available TLS signature algorithms?

### DISCUSSION THREAD

**Discussion by baentsch** [openssl/openssl#24522] [Q&A] [answered]
The [provider mechanism](https://www.openssl.org/docs/man3.3/man7/provider.html) allows (potentially many) different signature algorithms to be accessible to the `openssl` TLS logic. The [openssl config](https://www.openssl.org/docs/manmaster/man3/SSL_CONF_cmd.html) mechanism allows the ones actually active to be configured/limited, e.g., to [work around server problems with too many active signature algorithms](https://github.com/open-quantum-safe/oqs-provider/issues/399).

The question now is: Is there a mechanism by which the unwary user can retrieve a list of all permitted values to the "SignatureAlgorithms" config parameter, e.g., along the lines of `openssl list -signature-algorithms` (which does take providers into account, but does not output "fully qualified" algorithm names for use by the configuration, e.g. and incl. hash algs)?

**Comment by t8m** [comment]
There is clearly no such application currently. It would be an RFE to add it.

**Reply by baentsch** [reply]
Thanks, @t8m . Is there some openssl-specific info (documentation?) what an RFE (request for enhancement?) is and what (process) it would take to add this?

Second thought: If this is not _within_ the openssl code base, would it be conceivable that it is (existing or to be created) _outside_ of it? Something along the lines "run this script over your openssl installation and you'll get the list of parameters for SignatureAlgorithms"? (or any other config params that could be set)

**Reply by t8m** [reply, marked as answer]
Yes, request for enhancement. Just open a feature issue here in GH. Of course the prioritization of it is another matter.

**Comment by levitte** [comment]
`-signature-algorithms` lists all fetchable `EVP_SIGNATURE` implementations.  Those are signature algorithms, but not the kind you're looking for.  It might be good to be precise and talk about *TLS* signature algorithms (i.e. essentially the `TLS_SIGALG` capability that providers may have).

Could it be that what you're looking for is the output from `openssl ciphers`?  Or am I misunderstanding?

**Reply by baentsch** [reply]
> It might be good to be precise and talk about TLS signature algorithms (i.e. essentially the TLS_SIGALG capability that providers may have)

Correct. Title suitably changed.

> Could it be that what you're looking for is the output from openssl ciphers

The form of output is exactly what [we're looking for](https://github.com/open-quantum-safe/oqs-provider/issues/399#issuecomment-2119478364); the content isn't: It should be something along the lines of ```ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:ed25519:ed448:rsa_pss_pss_sha256:rsa_pss_pss_sha384:rsa_pss_pss_sha512:rsa_pss_rsae_sha256:rsa_pss_rsae_sha384:rsa_pss_rsae_sha512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:RSA+SHA224```.

Edit/add: And with something like oqsprovider enabled, it would become
```ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA512:ed25519:ed448:rsa_pss_pss_sha256:rsa_pss_pss_sha384:rsa_pss_pss_sha512:rsa_pss_rsae_sha256:rsa_pss_rsae_sha384:rsa_pss_rsae_sha512:RSA+SHA256:RSA+SHA384:RSA+SHA512:ECDSA+SHA224:RSA+SHA224:dilithium2:p256_dilithium2:rsa3072_dilithium2:dilithium3:p384_dilithium3:dilithium5:p521_dilithium5mldsa44:p256_mldsa44:rsa3072_mldsa44:mldsa44_pss2048:mldsa44_rsa2048:mldsa44_ed25519:mldsa44_p256:mldsa44_bp256:mldsa65:p384_mldsa65:mldsa65_pss3072:mldsa65_rsa3072:mldsa65_p256:mldsa65_bp256:mldsa65_ed25519:mldsa87:p521_mldsa87:mldsa87_p384:mldsa87_bp384:mldsa87_ed448falcon512:p256_falcon512:rsa3072_falcon512:falconpadded512:p256_falconpadded512:rsa3072_falconpadded512:falcon1024:p521_falcon1024:falconpadded1024:p521_falconpadded1024:sphincssha2128fsimple:p256_sphincssha2128fsimple:rsa3072_sphincssha2128fsimple:sphincssha2128ssimple:p256_sphincssha2128ssimple:rsa3072_sphincssha2128ssimple:sphincssha2192fsimple:p384_sphincssha2192fsimplesphincsshake128fsimple:p256_sphincsshake128fsimple:rsa3072_sphincsshake128fsimple```

**Reply by levitte** [reply]
Ah, I think I see what's going on here.  What you're saying looks like input and output are being conflated, with regards to what `openssl ciphers` does.

With `openssl ciphers`, an *input* like `ECDSA+SHA256` asks OpenSSL to find all TLS ciphersuites (which yeah, is a bit more than the mere signature algorithm) that match the input.  As an example, I get this with:

``` console
$ openssl ciphers ECDSA+SHA256
TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-ECDSA-CAMELLIA128-SHA256
```

or in verbose mode:

``` console
$ openssl ciphers -v ECDSA+SHA256
TLS_AES_256_GCM_SHA384         TLSv1.3 Kx=any      Au=any   Enc=AESGCM(256)            Mac=AEAD
TLS_CHACHA20_POLY1305_SHA256   TLSv1.3 Kx=any      Au=any   Enc=CHACHA20/POLY1305(256) Mac=AEAD
TLS_AES_128_GCM_SHA256         TLSv1.3 Kx=any      Au=any   Enc=AESGCM(128)            Mac=AEAD
ECDHE-ECDSA-AES128-SHA256      TLSv1.2 Kx=ECDH     Au=ECDSA Enc=AES(128)               Mac=SHA256
ECDHE-ECDSA-CAMELLIA128-SHA256 TLSv1.2 Kx=ECDH     Au=ECDSA Enc=Camellia(128)          Mac=SHA256
```

So it seems that you could kinda sorta get an output that answers what you're asking for, but at the same time, kinda not.

For your ask, `openssl ciphers` doesn't seem suitable after all.
Considering your ask, it would probably make sense to have the TLS signature algorithms listed with `-signature-algorithm`, *especially* if it's likely that they become directly fetchable as `EVP_SIGNATURE` in the future (when we finally get that support right).
